src/u2f.c

```diff
@@ -249,56 +330,104 @@ static int _sig_to_der(const uint8_t* sig, uint8_t* der)
  */
-static void _register(const USB_APDU* apdu, Packet* out_packet)
+static void _register(const Packet* in_packet, const USB_APDU* apdu, Packet* out_packet)
 {
-    uint8_t privkey[U2F_EC_KEY_SIZE] = {0};
-    uint8_t nonce[U2F_NONCE_LENGTH] = {0};
-    uint8_t mac[HMAC_SHA256_LEN] = {0};
-    uint8_t data[sizeof(U2F_REGISTER_RESP) + 2] = {0};
-    uint8_t sig[64] = {0};
-    U2F_REGISTER_SIG_STR sig_base;
-    U2F_REGISTER_RESP* response = (U2F_REGISTER_RESP*)data;
-    const U2F_REGISTER_REQ* reg_request = (const U2F_REGISTER_REQ*)apdu->data;
-
     if (APDU_LEN(*apdu) != sizeof(U2F_REGISTER_REQ)) {
         _error(U2F_SW_WRONG_LENGTH, out_packet);
+        U2F_USB_SEND(usb_processing_u2f(), out_packet);
         return;
     }
 
+    const U2F_REGISTER_REQ* reg_request = (const U2F_REGISTER_REQ*)apdu->data;
+    memcpy(&(_u2f_state.data.reg.reg_request), reg_request, sizeof(*reg_request));
+    _u2f_state.data.reg.cmd = in_packet->cmd;
+    _u2f_state.data.reg.cid = in_packet->cid;
+    _u2f_state.data.reg.apdu_p1 = apdu->p1;
+    _u2f_state.state = U2F_STATE_REGISTER;
+    _u2f_state.data.reg.state = U2F_STATE_REGISTRATION_REQUESTED;
+}
+
+/**
+ * Asks the user if he wants to proceed with registering.
+ */
+static void _perform_registration(void) {
+    const U2F_REGISTER_REQ* reg_request = &_u2f_state.data.reg.reg_request;
     // If the authentication fails with the "Bad key handle" the browser will execute bogus
     // registrations to make the device blink.
     bool is_bogus = MEMEQ(reg_request->appId, APPID_BOGUS_CHROMIUM, U2F_APPID_SIZE) ||
                     MEMEQ(reg_request->appId, APPID_BOGUS_FIREFOX, U2F_APPID_SIZE);
     if (!u2f_app_confirm(is_bogus ? U2F_APP_BOGUS : U2F_APP_REGISTER, reg_request->appId)) {
-        _error(U2F_SW_CONDITIONS_NOT_SATISFIED, out_packet);
+        _u2f_registration_failed(U2F_SW_CONDITIONS_NOT_SATISFIED);
         return;
     }
 
     if (!workflow_unlock()) {
-        _error(U2F_SW_CONDITIONS_NOT_SATISFIED, out_packet);
+        _u2f_registration_failed(U2F_SW_CONDITIONS_NOT_SATISFIED);
         return;
     }
+    _u2f_state.data.reg.state = U2F_STATE_REGISTRATION_COMPLETED;
+}
 
-    // Generate keys until a valid one is found
-    int i = 0;
-    for (;; ++i) {
-        if (i == 10) {
-            _error(U2F_SW_CONDITIONS_NOT_SATISFIED, out_packet);
-            return;
-        }
+/**
+ * Tries to generate a set of keys for signing a registration request.
+ *
+ * Fails after 10 attempts.
+ *
+ * @param[in] app_id
+ * @param[out] mac Buffer to fill with the generated message authentication code.
+ * @param[out] nonce Buffer to fill with the generated nonce.
+ * @param[out] pubkey Buffer to fill with the generated public key.
+ * @return Whether the generation was successful.
+ */
+static bool _generate_registration_keys(const uint8_t* app_id, uint8_t* nonce, uint8_t* mac, uint8_t* pubkey) {
+    uint8_t privkey[U2F_EC_KEY_SIZE];
+    for (int i = 0; i < 10; ++i) {
```

Comment:
can use `UTIL_CLEANUP_32(privkey)` on the next line and skip the manual zeroing.

Reply:
Old code, was refactored.